### PR TITLE
feat(connector): Add Revoke mandate flow

### PR DESCRIPTION
**Pull request by SamraatBansal**
## Type of Change

- [ ] Bugfix
- [X] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Add Revoke Mandate flow: When the merchant invokes the revoke-mandate endpoint, the mandate should be revoked at both the connector level and core level.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

- Create a mandate (zero and non-zero) with the connector
- Make Payment using the mandates
- Revoke the mandate
- Try to revoke an already revoked mandate

### Revoke a Mandate
Screenshot 2024-01-08 at 12 26 34 PM

### Try to revoke already revoked mandate
Screenshot 2024-01-08 at 12 26 45 PM

### Make Payment with revoked mandate
Screenshot 2024-01-08 at 12 27 18 PM

### Handle Error Response from connector
Screenshot 2024-01-08 at 3 28 17 PM

### Not implemented Error
Screenshot 2024-01-08 at 5 58 31 PM

## Checklist

- [X] I formatted the code `cargo +nightly fmt --all`
- [X] I addressed lints thrown by `cargo clippy`
- [X] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/core/mandate/utils.rs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+const IRRELEVANT_ATTEMPT_ID_IN_MANDATE_REVOKE_FLOW: &str =
+    "irrelevant_attempt_id_in_mandate_revoke_flow";
+
+const IRRELEVANT_CONNECTOR_REQUEST_REFERENCE_ID_IN_MANDATE_REVOKE_FLOW: &str =
+    "irrelevant_connector_request_reference_id_in_mandate_revoke_flow";
+
+pub async fn construct_mandate_revoke_router_data(
+    connector_name: &str,
+    request: mandates::MandateId,
```
Comment (Narayanbhat166) [inline crates/router/src/core/mandate/utils.rs:22]: `connector_name` is already present in `merchant_account`, and `mandate_id` is present in `mandate`, these parameters can be removed from the function definition

File: crates/router/src/core/mandate/utils.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        .change_context(errors::ApiErrorResponse::InternalServerError)?;
+    let router_data = types::RouterData {
+        flow: PhantomData,
+        merchant_id: merchant_account.merchant_id.clone(),
+        customer_id: Some(mandate.customer_id),
+        connector_customer: None,
+        connector: connector_name.to_string(),
+        payment_id: mandate
+            .original_payment_id
+            .unwrap_or_else(|| IRRELEVANT_PAYMENT_ID_IN_MANDATE_REVOKE_FLOW.to_string()),
```
Comment (Narayanbhat166) [inline crates/router/src/core/mandate/utils.rs:39]: Do we want to raise an error here if original_payment_id is not found @vspecky?

File: crates/router/src/core/mandate.rs [hunk 3]
```diff
@@ -44,26 +50,123 @@ pub async fn get_mandate(
...
+                            error_message: None,
+                        },
+                    ))
+                }
+
+                Err(err) => Ok(services::ApplicationResponse::Json(
+                    mandates::MandateRevokedResponse {
+                        mandate_id: mandate.mandate_id,
+                        status: mandate.mandate_status,
+                        error_code: None,
```
Comment (Narayanbhat166) [inline crates/router/src/core/mandate.rs:156]: do we not get `error_code` in response?

File: crates/router/src/core/mandate/utils.rs (hunk 2 above)
Comment (SamraatBansal) [reply crates/router/src/core/mandate/utils.rs:38]: There is a check in `revoke_mandate` function for this, if that check is not passed the pointer will not reach this point, so this is majorly a fallback to comply with the types and can be ignored

**Review by Narayanbhat166 — APPROVED** [review]

File: crates/router/src/connector/cybersource.rs [hunk 4]
```diff
@@ -434,7 +434,24 @@ impl
             })
         } else {
             // If http_code != 204 || http_code != 4xx, we dont know any other response scenario yet.
-            Err(errors::ConnectorError::ResponseHandlingFailed).into_report()
+            let response_value: serde_json::Value = serde_json::from_slice(&res.response)
+                .into_report()
+                .change_context(errors::ConnectorError::ResponseHandlingFailed)?;
+            let response_string: String = response_value
```
Comment (ArjunKarthik) [inline crates/router/src/connector/cybersource.rs:440]: let response_string = response_value.to_string();


Use to_string()

**Review by Narayanbhat166 — APPROVED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Pull request merged**